### PR TITLE
[MRG] Add viz of auto-noisy segment detection

### PR DESCRIPTION
**Pull request by hoechenberger**
This is currently non-interactive (purely static viz) until we've decided how to best add interactivity.

### CODE REVIEW

**Comment by agramfort** [comment]
Shall we aim to have this in the report ?

**Comment by hoechenberger** [comment]
> Shall we aim to have this in the report ?

Very good point. Yes I think we should. Need to adapt the pipeline to work with the new `BIDSPath` first, though, before proceeding with this PR!

File: 01-import_and_maxfilter.py [hunk 1]
```diff
@@ -97,6 +97,48 @@ def rename_events(raw, subject, session):
...
+
+        # Focus on scores that exceeded the limits.
+        sns.heatmap(data=data_to_plot,
+                    vmin=limits.min(), cmap='Reds',
+                    cbar_kws=dict(label='Score'), ax=ax[1])
+        ax[1].set_title('Scores > Limit', fontweight='bold')
+
+        # Figure title should not overlap with subplots.
+        fig.tight_layout(rect=[0, 0.03, 1, 0.95])
+        plt.show()
```
Comment (jasmainak) [inline 01-import_and_maxfilter.py:139]: why is this not in MNE?

**Comment by jasmainak** [comment]
can you share an example plot?

File: 01-import_and_maxfilter.py (hunk 1 above)
Comment (hoechenberger) [reply 01-import_and_maxfilter.py:139]: Hey @jasmainak, so the code in this PR is based on what can be found in the Maxwell filter tutorial of MNE, see https://mne.tools/dev/auto_tutorials/preprocessing/plot_60_maxwell_filtering_sss.html#using-sss-and-maxwell-filtering-in-mne-python

You can also find an example plot there.

The reason this is currently not in "MNE proper" is that we haven't decided how to best implement this. So I'd consider the current code in this PR as temporary, until we have a better solution built right into MNE!

**Comment by hoechenberger** [comment]
The figures are now automatically added to the report

Screenshot 2020-07-06 at 13 57 22

**Comment by agramfort** [comment]
@hoechenberger can you share  a rendered artifact on circle when you have one?

**Comment by hoechenberger** [comment]
@agramfort Absolutely!

… I see it just failed, need to fix that :(

**Comment by hoechenberger** [comment]
Also forgot to mention:
This now stores the scores returned by `find_bad_channels_maxwell()` as a JSON file, which is later used to retrieve those values when creating the figures.

**Comment by hoechenberger** [comment]
@agramfort 
https://511-167957354-gh.circle-artifacts.com/0/reports/ds000248/sub-01_task-audiovisual_report.html#6

**Comment by agramfort** [comment]
thx @hoechenberger 

cc @SophieHerbst

**Comment by hoechenberger** [comment]
Feedback I got from non-neuroscience friends who know their sh*it about design and accessibility:
- the gray vertical lines are awful
- the colormap may not be optimal; should try out "inferno", "plasma", "magma"
- interactivity typically adds a **huge** improvement in UX and usefulness to such figures

**Comment by agramfort** [comment]
no objection !

>

**Comment by SophieHerbst** [comment]
Very nice @hoechenberger! 
Do we now have a documentation somewhere of how the 'score' is computed (ideally in the bad channel detector function)?

**Comment by hoechenberger** [comment]
> Do we now have a documentation somewhere of how the 'score' is computed (ideally in the bad channel detector function)?


From the "Notes" section of the `find_bax_channels_maxwell()` doc:


Screenshot 2020-07-06 at 15 39 21

I tried to follow this through by stepping through the code line by line. I'm fine until including step 4; but step 5 either doesn't happen as described, or I'm misunderstanding something. What we do in the code in step 5, to my understanding, is we check if `zscore(dk) > limit`, where `limit` is user-supplied (and `7` by default).

What we return as the "noisy score" is `zscore(dk)`.

@larsoner could you shed some light onto this? I'm I missing something, or is the documentation outdated?

**Comment by hoechenberger** [comment]
Friendly ping to @larsoner – could you help me here? :)

**Comment by larsoner** [comment]
> I tried to follow this through by stepping through the code line by line. I'm fine until including step 4; but step 5 either doesn't happen as described, or I'm misunderstanding something. What we do in the code in step 5, to my understanding, is we check if zscore(dk) > limit, where limit is user-supplied (and 7 by default).

A `zscore(d)` is just `(d-mean)/std`. So if you're checking `d>mean+limit*std`, equivalently you're checking `(d-mean)/std>limit`, i.e. `zscore(d)>limit`

**Comment by hoechenberger** [comment]
> A `zscore(d)` is just `(d-mean)/std`. So if you're checking `d>mean+limit*std`, equivalently you're checking `(d-mean)/std>limit`, i.e. `zscore(d)>limit`

Bazinga! You're absolutely right, thank you!

Do you think it would make sense to adjust the user-facing documentation to make this relation between `limit` and `dk` more obvious? i.e., state that `d` is being z-transformed to a z-score, and that `limit` is given in z-scores too?

**Comment by larsoner** [comment]
A parenthetical mentioning that this is equivalent to the zscore exceeding limit would be fine

**Pull request merged**